File: mcdowell/python/ch9-04.py
def getAllSubSet(inputs, index):
    if index < 0:
        return [[]]

    ans = getAllSubSet(inputs, index - 1)
    newAns = ans[:]

    newInputValue = inputs[index]

    for v in ans:
        newV = v[:]
        newV.append(newInputValue)
        newAns.append(newV)

    return newAns



print getAllSubSet([1,2,3], 2)

File: mcdowell/js/ch9-04.js
function getAllSubSet(inputs, index) {
    if (index < 0) return [[]];

    var ans = getAllSubSet(inputs, index - 1);
    var newAns = ans.slice(0);

    var newInputItem = inputs[index];

    ans.forEach(function(v,k) {
        var newV = v.slice(0);
        newV.push(newInputItem);
        newAns.push(newV);
    });

    return newAns;
}


console.log(getAllSubSet([1,2,3], 2));

